java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/Server.java

```diff
@@ -0,0 +1,225 @@
+        File javaCluster = new File(jsClass.getProtectionDomain().getCodeSource().getLocation().toURI()).getParentFile().getParentFile();
+        System.setProperty("netbeans.dirs", javaCluster.getAbsolutePath());
+        CacheFolderProvider.getCacheFolderForRoot(Places.getUserDirectory().toURI().toURL(), EnumSet.noneOf(CacheFolderProvider.Kind.class), CacheFolderProvider.Mode.EXISTENT);
+
+        Lookup.getDefault().lookup(ModuleInfo.class); //start the module system
+
+        LanguageServerImpl server = new LanguageServerImpl();
+        Launcher<LanguageClient> serverLauncher = LSPLauncher.createServerLauncher(server, in, out);
+        ((LanguageClientAware) server).connect(serverLauncher.getRemoteProxy());
+        serverLauncher.startListening();
+
+        while (true) {
+            try {
+                Thread.sleep(100000);
+            } catch (InterruptedException ex) {
+                //ignore
+            }
+        }
+    }
+
+    private static class LanguageServerImpl implements LanguageServer, LanguageClientAware {
+
+        private static final Logger LOG = Logger.getLogger(LanguageServerImpl.class.getName());
+        private LanguageClient client;
+        private final TextDocumentService textDocumentService = new TextDocumentServiceImpl();
+
+        @Override
+        public CompletableFuture<InitializeResult> initialize(InitializeParams init) {
+            List<FileObject> projectCandidates = new ArrayList<>();
+            List<WorkspaceFolder> folders = init.getWorkspaceFolders();
+            if (folders != null) {
+                for (WorkspaceFolder w : folders) {
+                    try {
+                        projectCandidates.add(TextDocumentServiceImpl.fromUri(w.getUri()));
+                    } catch (MalformedURLException ex) {
+                        LOG.log(Level.FINE, null, ex);
+                    }
+                }
+            } else {
+                String root = init.getRootUri();
+
+                if (root != null) {
+                    try {
+                        projectCandidates.add(TextDocumentServiceImpl.fromUri(root));
+                    } catch (MalformedURLException ex) {
+                        LOG.log(Level.FINE, null, ex);
+                    }
+                } else {
+                    //TODO: use getRootPath()?
+                }
+            }
+            List<Project> projects = new ArrayList<>();
+            for (FileObject candidate : projectCandidates) {
+                Project prj = FileOwnerQuery.getOwner(candidate);
+                if (prj != null) {
+                    projects.add(prj);
+                }
+            }
+            //XXX: ensure project opened:
+            try {
+                Class.forName("org.netbeans.modules.project.ui.OpenProjectList", false, Lookup.getDefault().lookup(ClassLoader.class)).getDeclaredMethod("waitProjectsFullyOpen").invoke(null);
+            } catch (Exception ex) {
```

Comment:
Jackpot:
warning: Can be replaced with multicatch or several catch clauses catching specific exceptions